include/swift/Basic/Casting.h

```diff
@@ -0,0 +1,30 @@
+//===--- Casting.h - Helpers for casting ------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_CASTING_H
+#define SWIFT_BASIC_CASTING_H
+
+#include "STLCompatibility.h"
+
+namespace swift {
+
+/// Cast between two function types. Use in place of std::bit_cast, which
+/// doesn't work on ARM64e with address-discriminated signed function types.
```

Comment:
BTW, can you add some text on how/why the ptrauth parameter is dropped when invoking `swift::function_cast` but not `std::bit_cast`?

Reply:
I really wasn't thinking about this clearly. The actual problem with `bit_cast` is that it takes `source` by reference, which means we pick up ptrauth attributes on the field being passed in, if there are any. My version takes `source` by value, which avoids that problem. So that means that the existing `bit_cast` implementation works fine if we just do:

```
-bit_cast(const Source &src) noexcept {
+bit_cast(Source src) noexcept {
```

Would you prefer that instead, or would you prefer I keep this PR, implement `function_cast` in terms of `memcpy`, and add the additional commentary you suggest here?

Reply:
I think that we will need to have this wrapper even in the future then. The declaration for `std::bit_cast` is as follows:

```c++
template <class To, class From>
constexpr To bit_cast(const From& from) noexcept;
```

I think that adding the explanation and keeping `function_cast` makes sense then.

Reply:
Sounds good. Do you still want me to inline the implementation of `bit_cast` or just call the one from `STLCompatibility.h`? I don't have much of an opinion there, I'm happy either way.

Reply:
If there are sites where we continue to use `std::bit_cast` over `function_cast`, we shouldn't merge the headers.

Reply:
At the moment it’s only used for these function casts.

Reply:
Okay, then, lets go ahead and merge that into this header. The `std::bit_cast` is a compatibility shim until we can use C++20.